lib/Sema/CSDiagnostics.cpp

```diff
@@ -101,6 +101,52 @@ Expr *FailureDiagnostic::getArgumentExprFor(Expr *anchor) const {
   return nullptr;
 }
 
+ConstraintLocator *
+FailureDiagnostic::getConstraintLocatorFor(Expr *expr) const {
+  auto locator = CS.getConstraintLocator(expr);
+
+  if (isa<DeclRefExpr>(expr)) {
+    locator = CS.getConstraintLocator(expr);
+  } else if (isa<UnresolvedDeclRefExpr>(expr)) {
+    locator = CS.getConstraintLocator(expr);
+  } else if (isa<SubscriptExpr>(expr)) {
+    locator = CS.getConstraintLocator(expr, ConstraintLocator::SubscriptMember);
+  } else if (isa<UnresolvedDotExpr>(expr)) {
+    auto UDE = dyn_cast<UnresolvedDotExpr>(expr);
+    if (isa<ConstructorRefCallExpr>(UDE->getBase())) {
+      locator =
+          CS.getConstraintLocator(expr, ConstraintLocator::ConstructorMember);
+    } else {
+      locator = CS.getConstraintLocator(expr, ConstraintLocator::Member);
+    }
+  } else if (isa<MemberRefExpr>(expr)) {
+    auto MRE = dyn_cast<MemberRefExpr>(expr);
+    if (isa<ConstructorRefCallExpr>(MRE->getBase())) {
```

Comment:
Please use `getSelfForInitDelegationInConstructor(getDC(), expr)` here and in `UnresolvedDotExpr` case.